metricflow/test/table_snapshot/table_snapshots.py

```diff
@@ -0,0 +1,233 @@
+
+    # Pydantic feature to throw errors on extra fields.
+    class Config:  # noqa: D
+        extra = "forbid"
+
+    name: str
+    type: SqlTableColumnType
+
+
+class SqlTableSnapshotTypeException(Exception):  # noqa: D
+    pass
+
+
+class SqlTableSnapshot(FrozenBaseModel):
+    """Pydantic class to help parse table snapshots that are defined in YAML."""
+
+    # Pydantic feature to throw errors on extra fields.
+    class Config:  # noqa: D
+        extra = "forbid"
+
+    table_name: str
+    column_definitions: Tuple[SqlTableColumnDefinition, ...]
+    rows: Tuple[Tuple[Optional[str], ...], ...]
+    file_path: Path
+
+    @property
+    def snapshot_hash(self) -> SqlTableSnapshotHash:
+        """Return a hash that can be used to summarize the schema and data of the snapshot."""
+        return SqlTableSnapshotHash(
+            hash_items(
+                (self.table_name,)
+                + tuple(column_definition.name for column_definition in self.column_definitions)
+                + tuple(column_definition.type.name for column_definition in self.column_definitions)
+                + tuple(cell or "" for row in self.rows for cell in row)
+            )
+        )
+
+    @staticmethod
+    def _parse_bool_str(bool_str: str) -> bool:  # noqa: D
+        if bool_str.lower() == "false":
+            return False
+        elif bool_str.lower() == "true":
+            return True
+        else:
+            raise RuntimeError(f"Invalid string representation of a boolean: {bool_str}")
+
+    @property
+    def as_df(self) -> pd.DataFrame:  # noqa: D
+        """Return this snapshot as represented by an equivalent dataframe"""
+        # In the YAML files, all values are strings, but they need to be converted to defined type so that it can be
+        # properly represented in a dataframe
+
+        type_converted_rows = []
+        for row in self.rows:
+            type_converted_row: List[Union[str, datetime.datetime, int, float, bool, None]] = []
+            for column_num, column_value in enumerate(row):
+                column_type = self.column_definitions[column_num].type
+                if column_value is None:
+                    type_converted_row.append(None)
+                elif column_type is SqlTableColumnType.STRING:
+                    type_converted_row.append(column_value)
+                elif column_type is SqlTableColumnType.TIME:
+                    type_converted_row.append(dateutil.parser.parse(column_value))
+                elif column_type is SqlTableColumnType.INT:
+                    type_converted_row.append(int(column_value))
+                elif column_type is SqlTableColumnType.FLOAT:
+                    type_converted_row.append(float(column_value))
+                elif column_type is SqlTableColumnType.BOOLEAN:
+                    type_converted_row.append(SqlTableSnapshot._parse_bool_str(column_value))
```

Comment:
Urgh. I feel like we should be able to use pydantic union type collections for all of this and just have typed rows, but I guess that doesn't actually work. Sigh.

This is fine, I'm just complaining.